docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,436 @@
+type MachineSpec struct
+```
+- **To remove**
+    - **ProviderSpec [optional]** _Superseded by InfrastructureRef_
+        - Type: `*runtime.RawExtension`
+        - Description: ProviderSpec details Provider-specific configuration to use during node creation.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `MachineVersionInfo`
+        - Description: Versions of key software to use. This field is optional at cluster creation time, and omitting the field indicates that the cluster installation tool should select defaults for the user. These defaults may differ based on the cluster installer, but the tool should populate the values it uses when persisting Machine objects. A Machine spec missing this field at runtime is invalid.
+- **To add**
+    - **Bootstrap**
+        - Type: `Bootstrap`
+        - Description: Bootstrap is a reference to a local struct which encapsulates fields to configure the Machine’s bootstrapping mechanism.
+    - **InfrastructureRef [required]**
+        - Type: `corev1.TypedLocalObjectReference`
+        - Description: InfrastructureRef is a required reference to a Custom Resource Definition offered by an infrastructure provider.
+    - **Version [optional]**
+        - Type: `string`
+        - Description: Version defines the desired Kubernetes version. This field is meant to be optionally used by bootstrap providers.
+
+```go
+type MachineStatus struct
+```
+
+- **To remove**
+    - **ProviderStatus [optional]**
+        - Type: `*runtime.RawExtension`
+        - Description: ProviderStatus details a Provider-specific status. It is recommended that providers maintain their own versioned API types that should be serialized/deserialized from this field.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `*MachineVersionInfo`
+        - Description: Versions specifies the current versions of software on the corresponding Node (if it exists). This is provided for a few reasons [...].
+- **To add**
+    - **Version [optional]**
+        - Type: `*string`
+        - Description: Version specifies the current version of Kubernetes running on the corresponding Node. This is meant to be a means of bubbling up status from the Node to the Machine. It is entirely optional, but useful for end-user UX if it’s present.
+    - **Bootstrap [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the bootstrap provider.
+    - **Infrastructure [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the infrastructure provider.
+
+```go
+type Bootstrap struct
+```
+- **To add**
+    - **ConfigRef [optional]**
+        - Type: `*corev1.TypedLocalObjectReference`
+        - Description: ConfigRef is a reference to a bootstrap provider-specific resource that holds configuration details. The reference is optional to allow users/operators to specify Bootstrap.Data without the need of a controller.
+    - **Data [optional]**
```

Comment:
Should we be more explicit here instead of relying on special meanings for nil, empty, and non-empty?

Reply:
+1, we've been bitten by this in other apis.

Reply:
How would you suggest to be more explicit? Are you suggesting to add a better description or a different type?